packages/ai/core/generate-text/smooth-stream.ts

```diff
@@ -32,14 +41,48 @@ export function smoothStream<TOOLS extends ToolSet>({
 } = {}): (options: {
   tools: TOOLS;
 }) => TransformStream<TextStreamPart<TOOLS>, TextStreamPart<TOOLS>> {
-  const chunkingRegexp =
-    typeof chunking === 'string' ? CHUNKING_REGEXPS[chunking] : chunking;
+  let detectChunk: ChunkDetector;
 
-  if (chunkingRegexp == null) {
-    throw new InvalidArgumentError({
-      argument: 'chunking',
-      message: `Chunking must be "word" or "line" or a RegExp. Received: ${chunking}`,
-    });
+  if (typeof chunking === 'function') {
+    detectChunk = buffer => {
+      const match = chunking(buffer);
+
+      if (match == null) {
+        return null;
+      }
+
+      if (!match.length) {
+        throw new Error(`Chunking function must return a non-empty string.`);
+      }
+
+      if (!buffer.startsWith(match)) {
+        throw new Error(
+          `Chunking function must return a match that is a prefix of the buffer. Received: "${match}" expected to start with "${buffer}"`,
+        );
+      }
+
+      return match;
+    };
+  } else {
+    const chunkingRegex =
+      typeof chunking === 'string' ? CHUNKING_REGEXPS[chunking] : chunking;
+
+    if (chunkingRegex == null) {
+      throw new InvalidArgumentError({
+        argument: 'chunking',
+        message: `Chunking must be "word" or "line" or a RegExp. Received: ${chunking}`,
+      });
+    }
+
+    detectChunk = buffer => {
+      const match = chunkingRegex.exec(buffer);
+
+      if (!match) {
+        return null;
+      }
+
+      return buffer.slice(0, match.index) + match?.[0];
```

Comment:
prefer ternary with `match == null` condition (vs falsy which can match empty string

Reply:
aha correct

Reply:
actually wait no match is a regexarray object